# Making Recommendations Based on Popularity

These datasets are hosted on: https://archive.ics.uci.edu/ml/datasets/Restaurant+%26+consumer+data

They were originally published by: Blanca Vargas-Govea, Juan Gabriel González-Serna, Rafael Ponce-Medellín. Effects of relevant contextual features in the performance of a restaurant recommender system. In RecSys11: Workshop on Context Aware Recommender Systems (CARS-2011), Chicago, IL, USA, October 23, 2011.

## Restaurants data

In [1]:
import numpy as np
import pandas as pd

In [2]:
# rating_final.csv
url = 'https://drive.google.com/file/d/1ptu4AlEXO4qQ8GytxKHoeuS1y4l_zWkC/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
frame = pd.read_csv(path)

# chefmozcuisine.csv
url = 'https://drive.google.com/file/d/1S0_EGSRERIkSKW4D8xHPGZMqvlhuUzp1/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
cuisine = pd.read_csv(path)

# 'geoplaces2.csv'
url = 'https://drive.google.com/file/d/1ee3ib7LqGsMUksY68SD9yBItRvTFELxo/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
geodata = pd.read_csv(path, encoding = 'CP1252') # change encoding to 'mbcs' in Windows

On the "frame" dataset we have the ratings users have given to places. Ratings go from 0 to 2.

In [3]:
frame.head(50)

userID  placeID  rating  food_rating  service_rating
0   U1077   135085       2            2               2
1   U1077   135038       2            2               1
2   U1077   132825       2            2               2
3   U1077   135060       1            2               2
4   U1068   135104       1            1               2
5   U1068   132740       0            0               0
6   U1068   132663       1            1               1
7   U1068   132732       0            0               0
8   U1068   132630       1            1               1
9   U1067   132584       2            2               2
10  U1067   132733       1            1               1
11  U1067   132732       1            2               2
12  U1067   132630       1            0               1
13  U1067   135104       0            0               0
14  U1067   132560       1            0               0
15  U1103   132584       1            2               1
16  U1103   132732       0            0               2
17  U1103   132630       1            2               0
18  U1103   132613       2            2               2
19  U1103   132667       1            2               2
20  U1103   135104       1            2               0
21  U1103   132663       1            0               2
22  U1103   132733       2            2               2
23  U1107   132660       2            2               1
24  U1107   132584       2            2               2
25  U1107   132733       2            2               2
26  U1044   135088       2            2               2
27  U1044   132583       1            2               1
28  U1070   132608       2            2               1
29  U1070   132609       1            1               1
30  U1070   132613       1            1               0
31  U1031   132663       0            0               0
32  U1031   132665       0            0               0
33  U1031   132668       0            0               0
34  U1082   132630       1            1               1
35  U1023   132715       0            0               0
36  U1023   132717       0            0               0
37  U1082   132613       0            0               0
38  U1082   132594       0            0               0
39  U1082   132608       0            0               0
40  U1082   132609       0            0               0
41  U1082   132663       0            0               0
42  U1082   132560       0            0               0
43  U1060   132564       2            2               2
44  U1060   132717       2            2               2
45  U1060   132733       2            1               2
46  U1060   132740       1            2               1
47  U1015   135071       0            0               0
48  U1123   132733       1            2               1
49  U1123   132608       1            2               1

In the `geodata` dataset we have info about the places. We will only use the `name` column.

In [4]:
geodata

placeID   latitude   longitude  \
0     134999  18.915421  -99.184871   
1     132825  22.147392 -100.983092   
2     135106  22.149709 -100.976093   
3     132667  23.752697  -99.163359   
4     132613  23.752903  -99.165076   
..       ...        ...         ...   
125   132866  22.141220 -100.931311   
126   135072  22.149192 -101.002936   
127   135109  18.921785  -99.235350   
128   135019  18.875011  -99.159422   
129   132877  22.135364 -100.934948   

                                        the_geom_meter  \
0    0101000020957F000088568DE356715AC138C0A525FC46...   
1    0101000020957F00001AD016568C4858C1243261274BA5...   
2    0101000020957F0000649D6F21634858C119AE9BF528A3...   
3    0101000020957F00005D67BCDDED8157C1222A2DC8D84D...   
4    0101000020957F00008EBA2D06DC8157C194E03B7B504E...   
..                                                 ...   
125  0101000020957F000013871838EC4A58C1B5DF74F8E396...   
126  0101000020957F0000E7B79B1DB94758C1D29BC363D8AA...   
127  0101000020957F0000A6BF695F136F5AC1DADF87B20556...   
128  0101000020957F0000B49B2E5C6E785AC12F9D58435241...   
129  0101000020957F000090735015B84B58C1AF0DC0414698...   

                               name  \
0                   Kiku Cuernavaca   
1                   puesto de tacos   
2        El Rincón de San Francisco   
3    little pizza Emilio Portes Gil   
4                     carnitas_mata   
..                              ...   
125                         Chaires   
126                      Sushi Itto   
127                       Paniroles   
128     Restaurant Bar Coty y Pablo   
129                sirloin stockade   

                                             address             city  \
0                                         Revolucion       Cuernavaca   
1             esquina santos degollado y leon guzman           s.l.p.   
2                                    Universidad 169  San Luis Potosi   
3                            calle emilio portes gil        victoria    
4                             lic. Emilio portes gil         victoria   
..                                               ...              ...   
125                                 Ricardo B. Anaya  San Luis Potosi   
126               Venustiano Carranza 1809 C Polanco  San Luis Potosi   
127                                                ?                ?   
128  Paseo de Las Fuentes 24 Pedregal de Las Fuentes         Jiutepec   
129                                                ?                ?   

               state country fax  ...            alcohol   smoking_area  \
0            Morelos  Mexico   ?  ...  No_Alcohol_Served           none   
1             s.l.p.  mexico   ?  ...  No_Alcohol_Served           none   
2    San Luis Potosi  Mexico   ?  ...          Wine-Beer    only at bar   
3         tamaulipas       ?   ?  ...  No_Alcohol_Served           none   
4         Tamaulipas  Mexico   ?  ...  No_Alcohol_Served      permitted   
..               ...     ...  ..  ...                ...            ...   
125  San Luis Potosi  Mexico   ?  ...  No_Alcohol_Served  not permitted   
126              SLP  Mexico   ?  ...  No_Alcohol_Served           none   
127                ?       ?   ?  ...          Wine-Beer  not permitted   
128          Morelos  Mexico   ?  ...  No_Alcohol_Served           none   
129                ?       ?   ?  ...  No_Alcohol_Served           none   

    dress_code     accessibility   price                    url Rambience  \
0     informal  no_accessibility  medium  kikucuernavaca.com.mx  familiar   
1     informal        completely     low                      ?  familiar   
2     informal         partially  medium                      ?  familiar   
3     informal        completely     low                      ?  familiar   
4     informal        completely  medium                      ?  familiar   
..         ...               ...     ...                    ...       ...   
125   informal        completely  medium             

In [5]:
places =  geodata[['placeID', 'name']]
places.head()

placeID                            name
0   134999                 Kiku Cuernavaca
1   132825                 puesto de tacos
2   135106      El Rincón de San Francisco
3   132667  little pizza Emilio Portes Gil
4   132613                   carnitas_mata

In the `cuisine` dataset we have the type of cuisine that restaurants offer.

In [6]:
cuisine

placeID        Rcuisine
0     135110         Spanish
1     135109         Italian
2     135107  Latin_American
3     135106         Mexican
4     135105       Fast_Food
..       ...             ...
911   132005         Seafood
912   132004         Seafood
913   132003   International
914   132002         Seafood
915   132001   Dutch-Belgian

[916 rows x 2 columns]

## Popularity/Quality based recommmender system

Let's group places by rating, and look at their average rating. This is an **explicit** rating given by users.

In [7]:
frame.groupby('placeID')['rating'].mean()

placeID
132560    0.500000
132561    0.750000
132564    1.250000
132572    1.000000
132583    1.000000
            ...   
135088    1.000000
135104    0.857143
135106    1.200000
135108    1.181818
135109    1.000000
Name: rating, Length: 130, dtype: float64

In [8]:
rating = pd.DataFrame(frame.groupby('placeID')['rating'].mean())
rating.sort_values("rating", ascending=False).head()

rating
placeID          
132955   2.000000
135034   2.000000
134986   2.000000
132922   1.833333
132755   1.800000

The top rated places have a perfect score of 2/2. But how many reviews do these places have?

In [10]:
frame.query("placeID==132955")

userID  placeID  rating  food_rating  service_rating
934   U1004   132955       2            2               2
960   U1061   132955       2            2               2
996   U1059   132955       2            1               2
1014  U1097   132955       2            2               1
1080  U1096   132955       2            2               2

In [11]:
frame.query("placeID==135034")

userID  placeID  rating  food_rating  service_rating
61    U1083   135034       2            2               2
582   U1095   135034       2            2               2
721   U1048   135034       2            2               1
912   U1061   135034       2            2               2
1017  U1097   135034       2            2               1

Looks like only 5 people went to this place. Maybe they're just the owner's friends! Or maybe they're really top-quality places, but too niche to recommend to the masses.

We can also look at how many times each restaurant has received a rating. The ratings count is an **implicit** rating.

In [12]:
frame.groupby('placeID')['rating'].count()

placeID
132560     4
132561     4
132564     4
132572    15
132583     4
          ..
135088     6
135104     7
135106    10
135108    11
135109     4
Name: rating, Length: 130, dtype: int64

In [14]:
rating['rating_count'] = frame.groupby('placeID')['rating'].count()
rating.sort_values("rating_count", ascending=False).head()

rating  rating_count
placeID                        
135085   1.333333            36
132825   1.281250            32
135032   1.178571            28
135052   1.280000            25
132834   1.000000            25

Some places have been visited around 30 times. They are more popular than the top rated places, but received lower explicit ratings.

Let's locate the most popular place, and get some info about it:

In [15]:
rating.sort_values('rating_count', ascending=False).head()

rating  rating_count
placeID                        
135085   1.333333            36
132825   1.281250            32
135032   1.178571            28
135052   1.280000            25
132834   1.000000            25

In [16]:
rating.sort_values('rating_count', ascending=False).head(1)

rating  rating_count
placeID                        
135085   1.333333            36

In [17]:
rating.sort_values('rating_count', ascending=False).head(1).index[0]

135085

In [18]:
# placeId of most popular place
top_popular_placeID = rating.sort_values('rating_count', ascending=False).head(1).index[0]

# name of the most popular place
places[places['placeID']==top_popular_placeID]

placeID                    name
121   135085  Tortas Locas Hipocampo

In [19]:
# cuisine of the most popular place
cuisine[cuisine['placeID']==top_popular_placeID]

placeID   Rcuisine
44   135085  Fast_Food

The most popular place is "Tortas Locas Hipocampo", a fast food place that has received 36 reviews and it has an average score of 1.33.

### Challenge 1:

Find a hybrid system to sort restaurants, so that you can recommend the "best" places: restaurants that are both high rated and popular.

Create a new column that represents new ranking of both columns ratings and ratings_count e.g., rating * (ratings_count /5)

Join with correpsonding table to show the name of the resturent.


In [ ]:
# your code here

### Challenge 2:

Find a hybrid system to sort movies, so that you can recommend the "best" movies: movies that are both high rated and popular.

Join tables to show he name of the movies

In [ ]:
import pandas as pd

url = 'https://drive.google.com/file/d/1S0CtDB8NYUs94KgO0VDv6b2R1CShQcLF/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
links = pd.read_csv(path)


url = 'https://drive.google.com/file/d/1sW3zww6gMzoln0-U0Zs7HW_bKYjtH99i/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
movies = pd.read_csv(path)

url = 'https://drive.google.com/file/d/1nUpoWkhzhnYtUFvGYTR317RHiq7XtTx9/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
ratings = pd.read_csv(path)

url = 'https://drive.google.com/file/d/1F9szBIzHvE9sk-p89sk1zpxVEG_gJezg/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
tags = pd.read_csv(path)

In [ ]:
ratings

userId  movieId  rating   timestamp
0            1        1     4.0   964982703
1            1        3     4.0   964981247
2            1        6     4.0   964982224
3            1       47     5.0   964983815
4            1       50     5.0   964982931
...        ...      ...     ...         ...
100831     610   166534     4.0  1493848402
100832     610   168248     5.0  1493850091
100833     610   168250     5.0  1494273047
100834     610   168252     5.0  1493846352
100835     610   170875     3.0  1493846415

[100836 rows x 4 columns]